src/Microsoft.Tye.Hosting/Model/Application.cs

```diff
@@ -17,6 +18,18 @@ public Application(FileInfo source, Dictionary<string, Service> services)
             Source = source.FullName;
             ContextDirectory = source.DirectoryName!;
             Services = services;
+
+            foreach (var s in Services.Values)
+            {
+                if (s.Description.RunInfo is ProjectRunInfo projectRunInfo)
+                {
+                    var projectFileName = projectRunInfo.ProjectFile.FullName;
+                    if (!ProjectFileLocks.ContainsKey(projectFileName))
+                    {
+                        ProjectFileLocks[projectFileName] = new SemaphoreSlim(1, 1);
```

Comment:
Why can't these just be locks instead of semaphoreslims?

Reply:
Builds are called via [`ProcessUtil.RunAsync`](https://github.com/dotnet/tye/pull/895/files#diff-956d150a381630f9458461ccd954c3cf99e0be29aafd92b81a4ab98635f80763R372). You can't lock when making async calls.